warehouse/forklift/metadata.py

```diff
@@ -0,0 +1,352 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import email.message
+import email.utils
+import typing
+
+import email_validator
+
+from packaging.metadata import (
+    _LIST_FIELDS,
+    _RAW_TO_EMAIL_MAPPING,
+    _STRING_FIELDS,
+    InvalidMetadata,
+    Metadata,
+    RawMetadata,
+    _parse_keywords,
+    _parse_project_urls,
+)
+from packaging.requirements import InvalidRequirement, Requirement
+from trove_classifiers import all_classifiers, deprecated_classifiers
+from webob.multidict import MultiDict
+
+from warehouse.utils import http
+
+SUPPORTED_METADATA_VERSIONS = {"1.0", "1.1", "1.2", "2.0", "2.1"}
+
+
+# Mapping of fields on a Metadata instance to any limits on the length of that
+# field. Fields without a limit will naturally be unlimited in length.
+_LENGTH_LIMITS = {
+    "summary": 512,
+}
+
+
+class NoMetadataError(Exception):
+    pass
+
+
+def parse(
+    content: bytes | None, *, form_data: MultiDict | None = None, backfill: bool = False
+) -> Metadata:
+    # We prefer to parse metadata from the content, which will typically come
+    # from extracting a METADATA or PKG-INFO file from an artifact.
+    if content is not None:
+        metadata = Metadata.from_email(content)
```

Comment:
This method will raise if it encounters a non-standard field in the metadata. This can be a problem because it turns out that setuptools generates metadata that contains the not-yet-accepted `licence-file` field from PEP-639: https://github.com/pypa/packaging/issues/778